pkg/syncer/spec/spec_controller.go

```diff
@@ -39,39 +40,40 @@ import (
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/dynamic"
-	"k8s.io/client-go/dynamic/dynamicinformer"
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/kubernetes"
-	listersappsv1 "k8s.io/client-go/listers/apps/v1"
-	listerscorev1 "k8s.io/client-go/listers/core/v1"
-	listersrbacv1 "k8s.io/client-go/listers/rbac/v1"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/util/workqueue"
 	"k8s.io/klog/v2"
 
+	ddsif "github.com/kcp-dev/kcp/pkg/informer"
 	"github.com/kcp-dev/kcp/pkg/logging"
-	"github.com/kcp-dev/kcp/pkg/syncer/resourcesync"
+	"github.com/kcp-dev/kcp/pkg/syncer/indexers"
 	"github.com/kcp-dev/kcp/pkg/syncer/shared"
 	"github.com/kcp-dev/kcp/pkg/syncer/spec/dns"
 	specmutators "github.com/kcp-dev/kcp/pkg/syncer/spec/mutators"
 	. "github.com/kcp-dev/kcp/tmc/pkg/logging"
 )
 
 const (
-	controllerName              = "kcp-workload-syncer-spec"
-	byNamespaceLocatorIndexName = "syncer-spec-ByNamespaceLocator"
+	controllerName = "kcp-workload-syncer-spec"
 )
 
+var namespaceGVR schema.GroupVersionResource = corev1.SchemeGroupVersion.WithResource("namespaces")
+
 type Controller struct {
 	queue workqueue.RateLimitingInterface
 
 	mutators     mutatorGvrMap
 	dnsProcessor *dns.DNSProcessor
 
-	upstreamClient            kcpdynamic.ClusterInterface
-	downstreamClient          dynamic.Interface
-	syncerInformers           resourcesync.SyncerInformerFactory
-	downstreamNSInformer      informers.GenericInformer
+	upstreamClient   kcpdynamic.ClusterInterface
+	downstreamClient dynamic.Interface
+
+	getUpstreamLister                 func(gvr schema.GroupVersionResource) (kcpcache.GenericClusterLister, error)
+	getDownstreamLister               func(gvr schema.GroupVersionResource) (cache.GenericLister, error)
+	listDownstreamNamespacesByLocator func(jsonLocator string) ([]interface{}, error)
```

Comment:
why untyped?

Reply:
Didn't think at using the generic `indexers.ByIndex()`. I'll fix it.

Reply:
Fixed in commit da89d306186c852a5f9f9ab2a3a549f896f5f7e0